charms/tensorboard-controller/tests/unit/test_operator.py

```diff
@@ -0,0 +1,131 @@
+    ):
+        """Test that charm waits if it's not the leader."""
+        harness.begin_with_initial_hooks()
+
+        assert isinstance(harness.charm.model.unit.status, WaitingStatus)
+        assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_no_gateway_info_relation(
+        self, k8s_resource_handler: MagicMock, crd_resource_handler: MagicMock, harness: Harness
+    ):
+        """Test that charm waits if the gateway info relation is missing."""
+        harness.set_leader(True)
+        harness.begin_with_initial_hooks()
+
+        assert isinstance(harness.charm.model.unit.status, WaitingStatus)
+        assert harness.charm.model.unit.status == WaitingStatus(
+            "Waiting for gateway info relation"
+        )
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_active(
+        self, k8s_resource_handler: MagicMock, crd_resource_handler: MagicMock, harness: Harness
+    ):
+        """Test that charm goes to active status if the gateway info relation exists."""
+        harness.set_leader(True)
+        self._setup_gateway_info_relation(harness)
+
+        harness.begin_with_initial_hooks()
+
+        assert isinstance(harness.charm.model.unit.status, ActiveStatus)
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_pebble_layer(
+        self,
+        k8s_resource_handler: MagicMock,
+        crd_resource_handler: MagicMock,
+        harness: Harness,
+    ):
+        """Test creation of Pebble layer."""
+        harness.set_leader(True)
+        model, name = "kubeflow", "kubeflow-gateway"
+        harness.set_model_name(model)
+        self._setup_gateway_info_relation(harness, model, name)
+        harness.begin_with_initial_hooks()
+        harness.container_pebble_ready("tensorboard-controller")
+        assert harness.charm.container.get_service("tensorboard-controller").is_running()
+        pebble_plan = harness.get_container_pebble_plan("tensorboard-controller")
+        assert pebble_plan
+        assert pebble_plan.services
+        pebble_plan_info = pebble_plan.to_dict()
+        assert pebble_plan_info["services"]["tensorboard-controller"]["command"] == "/manager"
+        test_env = pebble_plan_info["services"]["tensorboard-controller"]["environment"]
+        # there should be 2 environment variables
+        assert 2 == len(test_env)
+        assert f"{model}/{name}" == test_env["ISTIO_GATEWAY"]
+
+    @patch("charm.TensorboardController.k8s_resource_handler")
+    @patch("charm.TensorboardController.crd_resource_handler")
+    def test_deploy_k8s_resources_success(
```

Comment:
This is good - what about testing the failure cases?  There's lots of logic in `_apply_k8s_resources()` (error catching, additional checks, etc)

Also, should we check on the `crd_resource_handler.apply()` as well?

Reply:
Most logic has been moved to the resource handler, which we mock here, so I think there's no need to go into any more detail.

Reply:
I'll assert that `crd_resource_handler.apply` is called as well, I hadn't noticed I skipped that.